rotkehlchen/chain/evm/decoding/hop/decoder.py

```diff
@@ -0,0 +1,311 @@
+
+    def _decode_transfer_from_l1(self, context: DecoderContext) -> DecodingOutput:
+        """This function is used to decode the TRANSFER_FROM_L1_COMPLETED event on Hop protocol."""
+        if not self.base.is_tracked(recipient := hex_or_bytes_to_address(context.tx_log.topics[1])):  # noqa: E501
+            return DEFAULT_DECODING_OUTPUT
+
+        if (bridge := self.bridges.get(context.tx_log.address)) is None:
+            return DEFAULT_DECODING_OUTPUT
+
+        for event in context.decoded_events:
+            if event.event_type == HistoryEventType.RECEIVE and event.event_subtype == HistoryEventSubType.NONE and recipient == event.location_label and event.asset.identifier == bridge.identifier:  # noqa: E501
+                event.event_type = HistoryEventType.WITHDRAWAL
+                event.event_subtype = HistoryEventSubType.BRIDGE
+                event.counterparty = CPT_HOP
+                event.notes = self._get_bridge_note(
+                    event.balance.amount,
+                    event.asset,
+                    recipient,
+                    string_to_evm_address(event.location_label) if event.location_label else None,
+                )
+                break
+
+        return DecodingOutput(matched_counterparty=CPT_HOP)
+
+    def _decode_token_swap(self, context: DecoderContext) -> DecodingOutput:
+        """Decodes a TokenSwap event to set the proper bridged amount"""
+        for item in context.action_items:
+            if item.to_event_subtype == HistoryEventSubType.BRIDGE:
+                tokens_bought = hex_or_bytes_to_int(context.tx_log.data[32:64])
+                asset = item.asset.resolve_to_evm_token()
+                amount = token_normalized_value(tokens_bought, asset)
+                item.to_notes = self._get_bridge_note(amount, asset)
+                break
+
+        return DEFAULT_DECODING_OUTPUT
+
+    def _decode_events(self, context: DecoderContext) -> DecodingOutput:
+        """This function is used to decode the bridging events done via Hop protocol."""
+        if context.tx_log.topics[0] == WITHDRAWAL_BONDED:
+            return self._decode_withdrawal_bonded(context=context)
+
+        if context.tx_log.topics[0] == TRANSFER_SENT:
+            return self._decode_transfer_sent(context=context)
+
+        if context.tx_log.topics[0] == WITHDREW:
+            return self._decode_withdrawal(context=context)
+
+        if context.tx_log.topics[0] == TRANSFER_FROM_L1_COMPLETED:
+            return self._decode_transfer_from_l1(context=context)
+
+        if context.tx_log.topics[0] == TOKEN_SWAP:
+            return self._decode_token_swap(context=context)
+
+        return DEFAULT_DECODING_OUTPUT
+
+    def addresses_to_decoders(self) -> dict[ChecksumEvmAddress, tuple[Any, ...]]:
+        addresses = {bridge.address for bridge in self.bridges.values()}
+        saddle_swaps = {bridge.saddle_swap for bridge in self.bridges.values() if bridge.saddle_swap}  # noqa: E501
```

Comment:
Question here: This seems too generic and all addresses pointing to the same handler. Then the handler checks topics and calls subhandlers in a big if elif else.

Can't you set different handlers per address or set of addresses? For example saddle swaps calling the decode token swap only?

I am not sure how the protocol works I am just asking since the decoder seems suboptimal at the moment with everything going into a single handler.

Reply:
Yes, setting the saddle swaps for tokenswap only makes sense. Updated.